### DISCUSSION TITLE
PKCE support documentation

### DISCUSSION THREAD

**Discussion by shrihari-prakash** [node-oauth/node-oauth2-server#196] [Q&A] [answered]
Is there any documentation available for enabling PKCE support that was added recently to this package?

**Comment by shrihari-prakash** [comment]
Hello @jankapunkt , if there is no documentation about this, is it possible to list the steps required to enable PKCE?

**Reply by jankapunkt** [reply, marked as answer]
Hi @shrihari-prakash sorry for the late response I am looking into this. I seems that we forgot to add a documentation.

The PKCE integration is based on [RFC 7636](https://datatracker.ietf.org/doc/html/rfc7636) and integrates with the authorization code grant workflow. The abstract workflow looks like the following:

```

                                                 +-------------------+
                                                 |   Authz Server    |
       +--------+                                | +---------------+ |
       |        |--(A)- Authorization Request ---->|               | |
       |        |       + t(code_verifier), t_m  | | Authorization | |
       |        |                                | |    Endpoint   | |
       |        |<-(B)---- Authorization Code -----|               | |
       |        |                                | +---------------+ |
       | Client |                                |                   |
       |        |                                | +---------------+ |
       |        |--(C)-- Access Token Request ---->|               | |
       |        |          + code_verifier       | |    Token      | |
       |        |                                | |   Endpoint    | |
       |        |<-(D)------ Access Token ---------|               | |
       +--------+                                | +---------------+ |
                                                 +-------------------+

                     Figure 2: Abstract Protocol Flow
```

>   A. The client creates and records a secret named the "code_verifier"
      and derives a transformed version "t(code_verifier)" (referred to
      as the "code_challenge"), which is sent in the OAuth 2.0
      Authorization Request along with the transformation method "t_m".

In this project the authorize endpoint usually calls `OAuth2Server.prototype.authorize` which itself uses `AuthorizeHandler`.
If your `Request` body contains `code_challenge` and `code_challenge_method` then PKCE is automatically included.

```js
const server = new OAuth2Server({ model })


// this could be added to express or other middleware
const authorizeEndpoint = function (req, res, next) {
  const request = new Request(req)
  server.authorize(request, response, options)
        .then(function (code) {
          // add code to response, code should contain 
        })
        .catch(function (err) {
          // handle error condition
        })
}
```

The request from the client must contain `code_challenge` and `code_challenge_method`.
As an example you could generate a random code challenge like so:

```js
const base64URLEncode = str => str.toString('base64')
      .replace(/\+/g, '-')
      .replace(/\//g, '_')
      .replace(/=/g, '')

// this is the code_verifier, which is kept secret on the client
// and which is later passed as request param to the token endpoint
const codeVerifier = base64URLEncode(crypto.randomBytes(32))

// this is the hashed version of the verifier, which is sent to the authorization endpoint
// this is named t(code_verifier) in the above workflow
const codeChallenge = base64URLEncode(crypto.createHash('sha256').update(codeVerifier).digest())

// this is the name of the code challenge method
// this is named t_m in the above workflow
const codeChallengeMethod = 'S256'

// add these to the request that is fired from the client
```

   B. The Authorization Endpoint responds as usual but records
      "t(code_verifier)" and the transformation method.

The `AuthorizeHandler.prototype.handle` saves code challenge and code challenge method automatically via `saveAuthorizationCode`. Note that this calls your model and you have to update your validation schema, if you use any.

   C. The client then sends the authorization code in the Access Token
      Request as usual but includes the "code_verifier" secret generated
      at (A).

This is usually done in your token endpoint, that should use `OAuth2Server.prototype.token`. 
Note that your client should have kept `codeVerifier` a secret until this step and now includes it
to the token endpoint.



   D. The authorization server transforms "code_verifier" and compares
      it to "t(code_verifier)" from (B).  Access is denied if they are
      not equal.

This is done by the package internally.


```js
const server = new OAuth2Server({ model })

// ...

// this could be added to express or other middleware
const tokenEndpoint = function (req, res, next) {
  const request = new Request(req)
  server.token(request, response, options)
        .then(function (code) {
          // add code to response, code should contain 
        })
        .catch(function (err) {
          // handle error condition
        })
}
```

**Comment by shrihari-prakash** [comment]
Hello @jankapunkt ,
Thanks much for the very detailed explanation. However, is there a change required in the method implementations in the model itself? For instance, the `saveAuthorizationCode` receives `code`, `client` and `user` as the input for the moment, in such cases should it also receive `codeChallenge` and `codeChallengeMethod`? If yes, what are the other methods that need to accept such new parameters in the model?

**Reply by jankapunkt** [reply]
hey @shrihari-prakash yes, I did not clearly described this but it's true. I only wrote `Note that this calls your model and you have to update your validation schema, if you use any.` What I wanted to say is that you have to update your `saveAuthorizationCode` implementation in order to accept  codeChallenge and codeChallengeMethod as arguments. From my current knowledge this should be the only necessary change in your model.

**Reply by shrihari-prakash** [reply]
It looks to me that the library is passing something else in the place of challenge and challenge method. This is what is printed in those variables:
![image](https://github.com/node-oauth/node-oauth2-server/assets/35889246/b9396ea4-f139-414b-9aa1-1a3acb3794bc)

For clarification, this is how my `saveAuthCode` looks:
https://github.com/shrihari-prakash/liquid/blob/8ced2876693c13f11084d1936ef6054895c9fea2/src/model/oauth.ts#L216

Am I doing something wrong?

**Reply by jankapunkt** [reply]
codeChallenge and codeChallengeMethod are attached to the `code` argument, they are not added as extra arguments.

**Reply by shrihari-prakash** [reply]
This seems to be everything I am getting with the code 🤔:
![image](https://github.com/node-oauth/node-oauth2-server/assets/35889246/6d493b65-421e-4fb9-bc9a-2af390b8d864)

Even though the request contains all these parameters:
![image](https://github.com/node-oauth/node-oauth2-server/assets/35889246/d167a3b3-fdf7-4ebb-bb75-4040fa2cd1a0)

**Reply by shrihari-prakash** [reply]
Oh so the OAuth Server seems to retrieve the code challenge and challenge method from the body, and returns nothing if it is in query string:
![image](https://github.com/node-oauth/node-oauth2-server/assets/35889246/5a0f9a53-024a-4e87-ad2d-06e81899beb4)

I suggest that we do something like:

```
return request.body.code_challenge || return request.query.code_challenge;;
```

Which would make it consistent with other parameters in the authorize method.

**Reply by shrihari-prakash** [reply]
For instance, see `getClient` method of https://github.com/node-oauth/node-oauth2-server/blob/44945641e488ed527a1fde2ba543a06dc2ed39ab/lib/handlers/authorize-handler.js#L168

**Reply by jankapunkt** [reply]
you are right, PR incoming

**Reply by jankapunkt** [reply]
Please review: https://github.com/node-oauth/node-oauth2-server/pull/197
